### ISSUE TITLE
Improve compilation performance (multithread?)

### ISSUE THREAD

**Issue by RSWilli** [go-gst/go-gst#16]
Currently the build takes a veeery long time. Although `htop` shows only minor usage over all cores.

Maybe this is just a configuration change needed to enable multithreading.

**Comment by RSWilli** [comment]
In combination with the LSP this regularily locks up my whole machine. "Minor usage" is only relevant for a single compilation.

See also https://github.com/golang/go/issues/50151#issuecomment-2304482522

**Comment by nickeygit** [comment]
> In combination with the LSP this regularily locks up my whole machine. "Minor usage" is only relevant for a single compilation.
> 
> See also golang/go#50151 (comment)

Hi Wilhelm,

I said much already abt tests (and still have in todo to show some),
but have an advice about this thing too.

In Clean Code (by Robert Martin / Uncle Bob), this situation is called viscosity.

If developers have to do many actions and/or wait for a long build/compile time to achieve a small thing - the system is viscose (moving like in honey).
Sure, this slows development down. And whole Clean Code is intended to fix the situation.

In our case, we have 2 issues here:
1. internal library functionality (all the locking/handles making etc) is hard-linked to another module - C gstreamer itself. Modifying anything in internal logic (which is NOT directly related to gstreamer) - we still have to recompile whole module together with the linkage layer to gstreamer. It's a pain.
The fix is - to split all the internal logic out to a separate package. So, C gstreamer functions calling would be in own pkg, like go-gst/bridge/. Users keep to use go-gst/gst as is. And it uses go-gst/bridge/ which contains actual cgo calls.
Or so.
So, modifying internal logic would not trigger C-linkages recompiling.

2. Big monolith. We have all the functionality (pretty different subparts of gstreamer!) in one big library. And sure, changing anything in it - triggers recompiling for it all.
Advice: is to split it down to smaller packages doing wrapping for different gstreamer subsystems: (pipeline, buses, gclosure etc). Then if one change a file - only its package would be recompiled. Nicely and fast.
For pure Go code this is not so important (as Go is fast! :D).
But for C-linkage to gstreamer - is significant.

So, having 2 directions of packages matrix:
* internal logic / c-linkage
* every gstreamer subsystem

can take us to pretty many separate packages - but will fix the viscosity issue.


PS I'd recommend to look "Clean Code" serial by Robert Martin.
First episode is free on youtube https://youtu.be/Wibk0IfjfaI (50 minutes)
Just absolutely wonderful information about why clean code matters.
(other episodes - on torrents :)

**Comment by RSWilli** [comment]
You are correct that smaller packages compile faster, that's one of the reasons for the few sub packages that are bindings for extensions/plugins of gstreamer.

I'm okay with this project compiling longer than usual go projects, but what bothers me is the way that the LSP/Formatter handles the project.

PS: interface segregation has nothing to to with package sizes/compile times.

**Comment by nickeygit** [comment]
> PS: interface segregation has nothing to to with package sizes/compile times.

If create proper interfaces which depend only on required set of code - then there will be no dependency on unused code. Including the unused code compile time. Not? :)

**Comment by RSWilli** [comment]
Would need purego -> a lot of work, so out of scope for now
